### PR TITLE
#1108 transaction status now uses single byte for higher TPS

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #1108  

Briefly describe the changes proposed in this PR:

* use single bytes for transaction status for higher TPS
* 
*

### CODE REVIEW

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java [hunk 1]
```diff
@@ -77,37 +82,48 @@ public void close() throws IOException {
...
+				status = TxnStatus.ROLLING_BACK;
+				break;
+			case 5:
+				status = TxnStatus.UNKNOWN;
+				break;
 			}
 		}
+
+		return status;
+
```
Comment (hmottestad) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: I should add a try catch or maybe a default case that uses the old strings based approach.

**Comment by hmottestad** [comment]
## Before
```
Benchmark                                    (numberOfTransactions)  Mode  Cnt      Score     Error  Units
TransactionsPerSecondBenchmark.transactions                     100  avgt   10     37.630 ±   1.527  ms/op
TransactionsPerSecondBenchmark.transactions                    1000  avgt   10    303.066 ±   2.984  ms/op
TransactionsPerSecondBenchmark.transactions                   10000  avgt   10   3171.710 ±  46.521  ms/op
TransactionsPerSecondBenchmark.transactions                  100000  avgt   10  36035.554 ± 368.616  ms/op

```

**Comment by hmottestad** [comment]
For transactions with 10 statements the difference is much much smaller. 1818 TPS vs 2028 TPS. 11% difference. I think that that's still significant enough to merge this.

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java [hunk 2]
```diff
@@ -13,41 +13,70 @@
...
+	// check that none of the new TxnStatus conflict with the old ones, ie. if the first byte for the old values are the
+	// same as the first byte for the new values
+	{
+		boolean assertsEnabled = TxnStatus.class.desiredAssertionStatus();
+		if (assertsEnabled) {
+			for (TxnStatus value1 : TxnStatus.values()) {
+				for (TxnStatus value2 : TxnStatus.values()) {
+					byte firstByteInOldValue = value1.name().getBytes(US_ASCII)[0];
+					byte firstByteInNewValue = value2.onDisk[0];
+					assert firstByteInOldValue != firstByteInNewValue;
```
Comment (jeenbroekstra) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: I don't quite understand what you're checking here. It looks as if it's testing that the new byte representations have no overlaps with "legacy" txn statuses. Why do we need to run that test as part of this code? Surely you can make sure of that once (the status enums are constants, after all)?

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java [hunk 3]
```diff
@@ -13,41 +13,70 @@
...
+
 /**
  * Writes transaction statuses to a file.
  */
 class TxnStatusFile {
+	public static final byte NONE_BYTE = (byte) 0b00000001;
+	public static final byte ACTIVE_BYTE = (byte) 0b00000010;
+	public static final byte COMMITTING_BYTE = (byte) 0b00000100;
+	public static final byte ROLLING_BACK_BYTE = (byte) 0b00001000;
+	public static final byte UNKNOWN_BYTE = (byte) 0b00010000;
```
Comment (jeenbroekstra) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: Can these be moved into the enum? And can they be made private?

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java [hunk 4]
```diff
@@ -13,41 +13,70 @@
...
-		ROLLING_BACK,
+		ROLLING_BACK(ROLLING_BACK_BYTE),
 
 		/**
 		 * The transaction status is unknown.
 		 */
-		UNKNOWN;
+		UNKNOWN(UNKNOWN_BYTE);
+
+		byte[] onDisk;
```
Comment (jeenbroekstra) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: Can we make this a private (final) field with a separate accessor method?

File: core/sail/nativerdf/src/test/java/org/eclipse/rdf4j/sail/nativerdf/NativeStoreTxnTest.java [hunk 5]
```diff
@@ -104,4 +109,35 @@ protected void removeStmt(Statement stmt) {
...
+		nioFile.truncate(bytes.length);
+		nioFile.writeBytes(bytes, 0);
+
+		repo.init();
+
+		try (RepositoryConnection con = repo.getConnection()) {
+
+			con.begin();
+			assertTrue(con.hasStatement(RDF.TYPE, RDF.TYPE, RDFS.RESOURCE, false));
+			con.commit();
```
Comment (jeenbroekstra) [inline core/sail/nativerdf/src/test/java/org/eclipse/rdf4j/sail/nativerdf/NativeStoreTxnTest.java]: Do you want to put a further assertion here that after replaying the txn and getting everything in shape, the transaction file now uses the new format?

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java (hunk 2 above)
Comment (hmottestad) [reply core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: I moved it into a test instead. It checks that the first byte in the old statuses isn't the same as the first byte in the new ones.

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java (hunk 3 above)
Comment (hmottestad) [reply core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: Done.

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java (hunk 4 above)
Comment (hmottestad) [reply core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: Done.

File: core/sail/nativerdf/src/test/java/org/eclipse/rdf4j/sail/nativerdf/NativeStoreTxnTest.java (hunk 5 above)
Comment (hmottestad) [reply core/sail/nativerdf/src/test/java/org/eclipse/rdf4j/sail/nativerdf/NativeStoreTxnTest.java]: I found the TripleStoreRecoveryTest and based my test on that one, with the exception that I hardcoded the old "COMMITTING" string.

**Comment by hmottestad** [comment]
Thanks for the great feedback @jeenbroekstra. Should be ready to roll now. What do you think? Should this be part of 3.1.0, or should we push it to 3.2.0?

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java [hunk 6]
```diff
@@ -77,25 +94,55 @@ public void close() throws IOException {
...
 	/**
 	 * Writes the specified transaction status to file.
-	 * 
+	 *
 	 * @param txnStatus The transaction status to write.
 	 * @throws IOException If the transaction status could not be written to file.
 	 */
 	public void setTxnStatus(TxnStatus txnStatus) throws IOException {
-		byte[] bytes = txnStatus.name().getBytes(US_ASCII);
+		byte[] bytes = txnStatus.onDisk;
```
Comment (jeenbroekstra) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: Does that work now that you've made it private? I would have expected you need to use `getOnDisk()` here instead.

**Comment by jeenbroekstra** [comment]
Btw happy to put this in for 3.1.0.

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java (hunk 6 above)
Comment (hmottestad) [reply core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/TxnStatusFile.java]: Was surprised too. Seems that this is legal. No compile issues.

**Pull request merged**